transport/grpc/outbound.go

```diff
@@ -234,3 +226,83 @@ func invokeErrorToYARPCError(err error, responseMD metadata.MD) error {
 	}
 	return intyarpcerrors.NewWithNamef(code, name, message)
 }
+
+// CallStream implements transport.StreamOutbound#CallStream.
+func (o *Outbound) CallStream(ctx context.Context, request *transport.StreamRequest) (*transport.ClientStream, error) {
+	if _, ok := ctx.Deadline(); !ok {
+		return nil, yarpcerrors.InvalidArgumentErrorf("stream requests require a connection establishment timeout on the passed in context.")
+	}
+	if err := o.once.WaitUntilRunning(ctx); err != nil {
+		return nil, err
+	}
+	return o.stream(ctx, request, time.Now())
+}
+
+func (o *Outbound) stream(
+	ctx context.Context,
+	req *transport.StreamRequest,
+	start time.Time,
+) (_ *transport.ClientStream, err error) {
+	if req.Meta == nil {
+		return nil, yarpcerrors.InvalidArgumentErrorf("stream request requires a request metadata")
+	}
+	treq := req.Meta.ToRequest()
+	md, err := transportRequestToMetadata(treq)
+	if err != nil {
+		return nil, err
+	}
+
+	fullMethod, err := procedureNameToFullMethod(req.Meta.Procedure)
+	if err != nil {
+		return nil, err
+	}
+
+	apiPeer, onFinish, err := o.peerChooser.Choose(ctx, treq)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { onFinish(err) }()
+
+	grpcPeer, ok := apiPeer.(*grpcPeer)
+	if !ok {
+		return nil, peer.ErrInvalidPeerConversion{
+			Peer:         apiPeer,
+			ExpectedType: "*grpcPeer",
+		}
+	}
+
+	tracer := o.t.options.tracer
+	createOpenTracingSpan := &transport.CreateOpenTracingSpan{
+		Tracer:        tracer,
+		TransportName: transportName,
+		StartTime:     start,
+		ExtraTags:     yarpc.OpentracingTags,
+	}
+	_, span := createOpenTracingSpan.Do(ctx, treq)
+
+	if err := tracer.Inject(span.Context(), opentracing.HTTPHeaders, mdReadWriter(md)); err != nil {
+		span.Finish()
+		return nil, err
+	}
+
+	streamCtx := metadata.NewOutgoingContext(context.Background(), md)
+	clientStream, err := grpcPeer.clientConn.NewStream(
+		streamCtx,
+		&grpc.StreamDesc{
+			ClientStreams: true,
+			ServerStreams: true,
+		},
```

Comment:
I don't know enough about GRPC to know the answer to this: Will we want to enable only one of these streams if the encoding-level APIs are unidirectional?

Reply:
Shouldn't need to.  We have needed to circumvent most of the requirements on the StreamDescs because we're using our own custom routing.  All the changes that these gate are resolvable without depending on the flags here.